tools/cldr-code/src/main/java/org/unicode/cldr/test/DisplayAndInputProcessor.java

```diff
@@ -1013,4 +998,98 @@ public static NumericType getNumericType(String xpath) {
             return posixDigitCount;
         }
     }
+
+    /**
+     * Turn all whitespace sequences (including tab and newline, and NBSP for certain paths)
+     * into a single space or a single NBSP depending on path.
+     * Also trim initial/final NBSP, unless the value is only the one character, "\u00A0"
+     *
+     * @param path
+     * @param value
+     * @return the normalized value
+     */
+    private String normalizeWhitespace(String path, String value) {
+        PathSpaceType pst = PathSpaceType.get(path);
+        if (pst == PathSpaceType.allowSp) {
+            value = WHITESPACE_AND_NBSP_TO_NORMALIZE.matcher(value).replaceAll(" "); // replace with regular space
+        } else if (pst == PathSpaceType.allowNbsp) {
+            value = WHITESPACE_AND_NBSP_TO_NORMALIZE.matcher(value).replaceAll("\u00A0"); // replace with NBSP
+            value = trimNBSP(value);
+        } else if (pst == PathSpaceType.allowSpOrNbsp) {
+            /*
+             * in this case don't normalize away NBSP
+             */
+            value = WHITESPACE_NO_NBSP_TO_NORMALIZE.matcher(value).replaceAll(" "); // replace with regular space
```

Comment:
Not sure this does what we want.

I would do it with:

ALL_WHITESPACE = [\\s\\u00A0]+
matches all sequences of whitespace and NBSP

ALL_WHITESPACE_MUST_HAVE_NBSP = \\s*\\u00A0[\\s\\u00A0]* 
matches all sequences of whitespace and NBSP, AND has at least one NBSP in it.

Then for the allowSpOrNbsp, 

1. convert ALL_WHITESPACE_MUST_HAVE_NBSP to NBSP
2. then (anything with NBSP having been handled) do
3. convert ALL_WHITESPACE to SPACE

Your 
NBSP_PLUS_SPACE_TO_NORMALIZE
SPACE_PLUS_NBSP_TO_NORMALIZE
doesn't really do this, because it requires NBSP to be at the start or end. But we allow interior NBSP also.

Reply:
"requires NBSP to be at the start or end" -- I think in combination this code already handles interior NBSP, as shown by the unit tests, which do pass. Can you think of a test that would fail? Anyway I'll study it further.

Reply:
The unit test shows that for an `allowSpOrNbsp` path, "\u00A0  X \u00A0 Y \u00A0" is correctly converted to "X\u00A0Y". The interior NBSP has ordinary spaces before and after it. The space after it gets removed by NBSP_PLUS_SPACE_TO_NORMALIZE, and the space before it gets removed by SPACE_PLUS_NBSP_TO_NORMALIZE.